fix: throw on failed bytestring parsing

nomadalgia opened the pull request:
This closes internal issue #1706
The functional application order is sometimes different but I just stuck to what was already there.
I hope the solutions aren't too pointfree

Original issue summary:

Currently, Database.Redis.Schema joins the Maybe from `fromBS` and get into a single one; meaning that if something fails in Serialization, it is observed as if the key doesn't exists. We currently know this `readBinary` can (and does) fail on some entries. We don't know which shape these have. But at least the current form of [...] should work on some entries as well? At least prior all failed.

nomadalgia committed:
fix: throw on failed bytestring parsing

nomadalgia committed:
fix: cover some other cases of fromBS within RedisM

nomadalgia committed:
fix: cover the transaction gets as well

The txCheckMap solution was modelled based on txFromBSMany

cdfa commented on src/Database/Redis/Schema.hs:
```diff
@@ -400,6 +400,16 @@ class Value inst val where
   valSetTTLIfExists (SviHash keyBS _hkeyBS) (TTLSec ttlSec) =
     expectRight "valSetTTLIfExists/hash" =<< Hedis.expire keyBS ttlSec
 
+-- | Ensure the fromBS conversation does not fail silently due to a change of the internal Haskell ValueType of a Ref
```
conversation -> conversion

nomadalgia replied:
Whoopsie, thanks!

cdfa commented on src/Database/Redis/Schema.hs:
```diff
@@ -368,9 +368,9 @@ class Value inst val where
   default valGet :: SimpleValue inst val => Identifier val -> RedisM inst (Maybe val)
   valGet (SviTopLevel keyBS) =
-    fmap (fromBS =<<) . expectRight "valGet/plain" =<< Hedis.get keyBS
+    (traverse rFromBSorThrow <=< expectRight "valGet/plain") =<< Hedis.get keyBS
```
I find `traverse rFromBSorThrow =<< expectRight "valGet/plain" =<< Hedis.get keyBS` cleaner personally

nomadalgia replied:
Me too :x

cdfa commented on src/Database/Redis/Schema.hs:
```diff
@@ -400,6 +400,16 @@ class Value inst val where
+rFromBSorThrow bs = maybe (throw $ CouldNotDecodeValue $ Just bs) pure $ fromBS bs
+
+txFromBSorExcept :: Serializable a => Tx inst (Maybe ByteString) -> Tx inst (Maybe a)
+txFromBSorExcept = txCheckMap $ traverse $ maybeToEither . CouldNotDecodeValue . Just <*> fromBS
```
So the `<*>` uses the `(->) Bytestring` functor? That's unusual enough that I had to go inspect the instantiation of `<*>` myself. I'm a fan of pointfree code in general as well, but here I would prefer `txCheckMap $ traverse $ \bs -> maybe (Left $ CouldNotDecodeValue $ Just bs) Right $ fromBS bs`

nomadalgia replied:
Yeah, because of `(->)`'s Applicative definition, `(<*>)` is actually the S combinator from combinatory logic. [`λabc.ac(bc)`]
I'm myself quite used to most of the combinators found [here](https://combinatorylogic.com/table.html)... I tend to forget they're quite obfuscating =D
I've changed it to your suggestion!

cdfa replied:
Ahh I didn't know! That makes it easier to understand actually :)

cdfa commented:
Looks good functionality wise, though I do have some style comments.
I also discovered Matus already created a [branch](https://github.com/chordify/redis-schema/commits/report-parsing-errors) doing the same thing upstream. You might one to ping him about which to merge.
Also, you don't have to create your own fork of our repositories as employee :).
I'm sure someone in backend can add you to our github org.

nomadalgia committed:
refactor: process review

Review: https://github.com/chordify/redis-schema/pull/10#issuecomment-3562995999
Changes:
	* typo: conversation > conversion
	* use =<< instead of parenthetical <=<
	* replace S combinator with lambda abstraction

nomadalgia commented:
@cdfa 
>Looks good functionality wise, though I do have some style comments.

Addressed! (hopefully)

> I also discovered Matus already created a [branch](https://github.com/chordify/redis-schema/commits/report-parsing-errors) doing the same thing upstream. You might one to ping him about which to merge.

Oh wow what o.O
He gave me this issue to work on; curious.
I'll ask him about it... I didn't even look at the open PR's here before starting :x

>  Also, you don't have to create your own fork of our repositories as employee :).
> I'm sure someone in backend can add you to our github org.

I'm gonna ping the open sourcerer, thanks.
It seemed like not everyone was in the GitHub org, so I thought it's not for everyone

cdfa reviewed (approved):
LGTM! Thanks @nomadalgia !
